### ISSUE TITLE
Future deviations in operating model

### ISSUE THREAD

**Issue by k-doering-NOAA** [nmfs-fish-tools/SSMSE#45]
The last large scale feature we want to develop for SSMSE is a way to allow the operating model to change in the future (rather than just carrying forth the assumptions in the initial model). For example, we may want selectivity to change for a fleet in some year in the future. Another example would be allowing natural mortality to deviate from year to year or have a trend.

Hopefully, we can develop a unified approach to allowing users to add change to their operating model. Ideally, they could pass a dataframe  that describes when and what changes could be made.

@nathanvaughan-NOAA , feel free to edit or add to this comment. I'm thinking we can communicate about our ideas re: implementations on this thread.

**Comment by nathanvaughan-NOAA** [comment]
That sounds great to me

**Comment by kellijohnson-NOAA** [comment]
Would it be more robust / easier to allow users to provide a new ctl file for the OM for different years? Where essentially right now you have one OM ctl for all years. Users could specify a vector of ctl files, each year you could point to a file, and if you only have one file it will be applied to all years? Or, something along those lines. That allows users to change the file in all kinds of ways.

**Comment by nathanvaughan-NOAA** [comment]
That could be a good idea as an advanced option. I think that would be too
complex though as a general approach for simple things like a random walk
on a parameter like M or selectivity.

**Comment by k-doering-NOAA** [comment]
TLDR: I agree with Nathan!

Hmm, I hadn't thought of doing it that way.  To me, I think it seems easier to keep track of a single control file for an operating model. Like we found out with ss3sim, specifying inputs in a dataframe vs. in several files is more manageable. 
Also, I'm not sure if the user would feel confident doing this for an OM that is getting feedback from the estimation method and being extended forward in time? To me, it seems like we could get some strange bugs? But maybe I am not thinking about it correctly

I guess I was thinking we could allow users to specify which variable they want to change, then have several options for how they want to change it. I see how this would have limitations, though; if someone wanted to make large future structural changes, this would be difficult.

I wonder if there is some way we could make it so the coding would be easy for simple changes but also accommodate more complex ones like having multiple control files would allow?

**Comment by kellijohnson-NOAA** [comment]
We struggled with this as well with ss3sim and didn't find a good way. For example, if users want to estimate growth outside of the model how would they input those values into the EM? Given you are only talking about the OM here, where all OMs are SS models, I think your original proposal is better than what I was thinking given that you have read and write capabilities of the ctl file. Where it gets stickier is in the EM b/c you allow multiple frameworks.

**Comment by k-doering-NOAA** [comment]
Yeah, I see how the EM could be more difficult, thanks for explaining, Kelli! 

We recently added some code to allow folks to [write their own function for the management procedure](https://nmfs-fish-tools.github.io/SSMSE/manual/SSMSE.html#using-a-custom-management-strategyprocedure) (including the estimation method), so in theory, the user could write an R function to handle their own custom way of manipulating the EM; this should work as long as they can use the data coming out of the OM (which could also be manipulated before).

So, I think this issue will just be left as specific to the OM, rather than EM, for now.

**Comment by k-doering-NOAA** [comment]
Some futher notes (temporarily) in the wiki: https://github.com/nmfs-fish-tools/SSMSE/wiki/Notes-for-future-deviations-in-OM-design

Another aspect of this will be dealing with time varying parameters that are already in the model.

**Comment by k-doering-NOAA** [comment]
Finished and merged to main
